[browser] [wasm] Refactor Request Streaming to use HttpContent.CopyToAsync

campersau opened the pull request:
Uses a [`TransformStream`](https://developer.mozilla.org/en-US/docs/Web/API/TransformStream) into which the `HttpContent` gets [copied into](https://developer.mozilla.org/en-US/docs/Web/API/WritableStreamDefaultWriter). Sync writes are not allowed like sync reads in response streaming.

Ref: https://github.com/dotnet/runtime/pull/91295#issuecomment-1706919710

campersau committed:
try out transformstream

pavelsavara commented on src/libraries/Common/tests/System/Net/Http/ResponseStreamTest.cs:
```diff
@@ -307,15 +310,16 @@ public async Task BrowserHttpHandler_StreamingRequest()
             {
                 Assert.Equal(HttpStatusCode.OK, response.StatusCode);
-                Assert.Equal(size.ToString(), Assert.Single(response.Headers.GetValues("X-HttpRequest-Body-Length")));
+                Assert.Equal((size + 129).ToString(), Assert.Single(response.Headers.GetValues("X-HttpRequest-Body-Length")));
```
Could you please explain +129 ? Maybe made it a named constant.

campersau replied:
This test now sends `MultipartFormDataContent` which adds some more things like boundaries, field name ("test" in this case) and multi part headers. Will add a constant for it.

pavelsavara commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -65,80 +65,32 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
+    // the bufferPtr is pinned by the caller
+    const view = new Span(bufferPtr, bufferLength, MemoryViewType.Byte);
+    const copy = view.slice() as Uint8Array;
+    await ts.__writer.write(copy);
```
I would love this to be cancellable.

pavelsavara replied:
I assume this may throw when
- TPC breaks or server kills the connection mid flight
- we triggered AbortController

campersau replied:
Unfortunately not, I need to race all of these with the `fetch_promise_control` e.g. 
```js
await Promise.race([ts.__writer.write(copy), ts.__fetch_promise_control.promise]);
```
or do you have another idea?

```js
var t = new TransformStream();
var w = t.writable.getWriter();
var a = new AbortController();

var f = fetch("", {
    method: "POST",
    body: t.readable,
    duplex: "half",
    signal: a.signal
});

a.abort();

var wp = w.write(new Uint8Array([1, 2, 3]));

var done = Promise.race([wp, f]);

// wp is pending
// done / f is rejected
```

campersau replied:
I opened a bug report in chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1480250
Because I would expect that any fetch error would properly cancel the `ReadableStream` (which the browser locks) and thus any writer promises would be rejected.

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpInterop.cs:
```diff
@@ -28,16 +28,16 @@ internal static partial class BrowserHttpInterop
-            [JSMarshalAs<JSType.Any>] object pullState,
-            Exception error);
+        [JSImport("INTERNAL.http_wasm_request_stream_close")]
+        public static partial Task<JSObject> RequestStreamClose(
```
This confuses me, could you please explain why we are returning promise here ?

campersau replied:
This returns the actual `fetchResponse` promise.

So it works like this:

1. The `FetchStream` method returns the `TransformStream`.
2. `CopyToAsync` to the `TransformStream`.
3. Close the `TransformStream` and return the actual `fetchResponse` to read the response from.

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -329,41 +318,98 @@ static async Task<HttpResponseMessage> Impl(HttpRequestMessage request, Cancella
+                await RequestStreamWriteUnsafe(_requestStream, buffer, handle).ConfigureAwait(true);
             }
+
+            static unsafe Task RequestStreamWriteUnsafe(JSObject requestStream, ReadOnlyMemory<byte> buffer, Buffers.MemoryHandle handle)
```
Please make it cancellable promise. The pattern is already in the previous code. Or I can try to help :)

pavelsavara replied:
When this propagates cancelation exception, or if this was aborted,
how do we propagate that error to the caller of the top level `.SendAsync`, does that work fine ?

campersau replied:
I am planning to do it like this https://github.com/dotnet/runtime/pull/91699#discussion_r1319082854 so it should be automatically cancellable since fetch is cancellable.

pavelsavara reviewed (commented):
Nice work thanks! 
I wonder if the transformer could also improve response streaming :)

campersau committed:
PR feedback

campersau committed:
race transform stream writing with fetch promise

unfortunately the transform stream does not get notified when fetch fails or cancels so we need to race with it

campersau commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -65,79 +65,36 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
-
-    function cancel(error: any) {
-        pull_state.__fetch_promise_control?.reject(error);
+export async function http_wasm_transform_stream_close(ts: TransformStreamExtension, error: Error | undefined): Promise<void> {
```
This could be split into two methods, but it currently saves one export/import.

pavelsavara replied:
Let's split it, for clarity sake

campersau commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -329,41 +326,103 @@ static async Task<HttpResponseMessage> Impl(HttpRequestMessage request, Cancella
+                try
+                {
+                    Task copyTask = content.CopyToAsync(stream, cancellationToken);
+                    if (await Task.WhenAny(copyTask, fetchResponsePromise).ConfigureAwait(true) == copyTask)
```
Race with fetch to get errors and cancellation because TransformStream does not propagate these.

pavelsavara replied:
I wonder if race on TS side could propagate the failure to copyTask, so that we don't have to race here.

campersau replied:
It depends on what the behavior should be when a request content just waits indefinitely in `SerializeToStreamAsync`. E.g. should it throw a cancellation exception or not.

I checked the normal `SocketsHttpHandler` and it waits indefinitely so this can be removed.

```c#
var cts = new CancellationTokenSource();
cts.CancelAfter(TimeSpan.FromSeconds(5));

using var client = new HttpClient();
client.Timeout = TimeSpan.FromSeconds(15);
await client.PostAsync("https://github.com", new MyHttpContent(), cts.Token); // this will wait indefinitely

class MyHttpContent : HttpContent
{
    protected override Task SerializeToStreamAsync(Stream stream, TransportContext? context)
        => new TaskCompletionSource().Task;

    protected override bool TryComputeLength(out long length)
    {
        length = 0;
        return false;
    }
}
```

campersau commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -65,79 +65,36 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
+export async function http_wasm_transform_stream_write(ts: TransformStreamExtension, bufferPtr: VoidPtr, bufferLength: number): Promise<void> {
+    mono_assert(bufferLength > 0, "expected bufferLength > 0");
+    mono_assert(ts.__fetch_promise_control, "expected fetch promise control");
+    await Promise.race([ts.__writer.ready, ts.__fetch_promise_control.promise]);
```
Race with fetch everywhere to get errors and cancellation because TransformStream does not propagate these.

pavelsavara replied:
If fetch failed/aborted/whatever it would throw and we could catch it here ? And abort the other promise via it's controller too ?

pavelsavara replied:
transformer ends have `cancel()` and `abort()`

pavelsavara replied:
We could still wrap those promises with cancelable and have our own controller

pavelsavara replied:
`ts.__writer.abort(error);` is not locked after all ?

Anyway, that solves the case when we are aborting it. I think the case when TCP fails is still hanging ?

campersau replied:
> ts.__writer.abort(error); is not locked after all ?

Everything on the writer is not locked because we created it.
Only the `ReadableStream` is locked because the browser is using it.

I tried forwarding fetch errors to `writer.abort(error)` but it doesn't work in all cases. E.g.

```js
var t = new TransformStream();
var w = t.writable.getWriter();
var a = new AbortController();
var f = fetch("", {
    method: "POST",
    body: t.readable,
    duplex: "half",
    signal: a.signal
});
f.catch(error => w.abort(error));

var wp = w.ready.then(() => { a.abort(); return w.write(new Uint8Array([1, 2, 3])); });

// wp is pending and never resolves
// f will be rejected
```

The `BrowserHttpHandler_StreamingRequest_Http1Fails` would fail because the `write()` promise never resolves as fetch gets rejected during that write...

campersau committed:
add more tests for http1 failed fetch and cancellation

campersau committed:
PR feedback

campersau committed:
test more return types for request cancellation

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -220,12 +220,17 @@ private static async Task<WasmFetchResponse> CallFetch(HttpRequestMessage reques
+                        JSObject transformStream = BrowserHttpInterop.CreateTransformStream();
+                        try
+                        {
+                            promise = BrowserHttpInterop.Fetch(uri, headerNames.ToArray(), headerValues.ToArray(), optionNames, optionValues, abortController, transformStream);
```
let's rename this one to fetchPromise or something.

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -220,12 +220,17 @@ private static async Task<WasmFetchResponse> CallFetch(HttpRequestMessage reques
+                        try
+                        {
+                            promise = BrowserHttpInterop.Fetch(uri, headerNames.ToArray(), headerValues.ToArray(), optionNames, optionValues, abortController, transformStream);
+                            promise = WasmHttpWriteStream.CopyToAsync(promise, transformStream, request.Content, cancellationToken);
```
this second promise is not a proxy of the JS promise and so, it would not work with `BrowserHttpInterop.CancelationHelper`, it will assert/throw I think.

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -329,41 +326,99 @@ static async Task<HttpResponseMessage> Impl(HttpRequestMessage request, Cancella
+                {
+                    await content.CopyToAsync(stream, cancellationToken).ConfigureAwait(true);
+                    await BrowserHttpInterop.TransformStreamClose(transformStream).ConfigureAwait(true);
+                    return await fetchResponsePromise.ConfigureAwait(true);
```
could this await move back up to CallFetch ?

campersau replied:
Yes, I have rearranged the code.

pavelsavara commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -65,79 +65,37 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
+    // the bufferPtr is pinned by the caller
+    const view = new Span(bufferPtr, bufferLength, MemoryViewType.Byte);
+    const copy = view.slice() as Uint8Array;
+    await Promise.race([ts.__writer.write(copy), ts.__fetch_promise_control.promise]);
```
wrap this promise with 
```
return wrap_as_cancelable_promise(async () => {
        return Promise.race([ts.__writer.write(copy), ts.__fetch_promise_control.promise])
    });
```

and make the method synchronous (on JS side only)

pavelsavara replied:
So that `async` doesn't create new promise (without the controller on it)

campersau committed:
PR feedback wrap_as_cancelable_promise

campersau committed:
PR feedback rename to fetchPromise

campersau committed:
test async cancel

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -329,41 +336,81 @@ static async Task<HttpResponseMessage> Impl(HttpRequestMessage request, Cancella
+            using (Buffers.MemoryHandle handle = buffer.Pin())
             {
-                BrowserHttpInterop.ReadableStreamControllerError(this, ex);
+                await TransformStreamWriteUnsafe(_transformStream, buffer, handle).ConfigureAwait(true);
```
add `BrowserHttpInterop.CancelationHelper`

campersau replied:
Yes, currently working on it :)

pavelsavara replied:
Because also C# async creates new Task instances

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -220,28 +220,43 @@ private static async Task<WasmFetchResponse> CallFetch(HttpRequestMessage reques
+                                try
+                                {
+                                    await request.Content.CopyToAsync(stream, cancellationToken).ConfigureAwait(true);
+                                    await BrowserHttpInterop.TransformStreamClose(transformStream).ConfigureAwait(true);
```
does close need to be cancelable too ? could it take long time to finish pusing bytes to network in the stage ?

campersau replied:
Yes, it does.

campersau committed:
use CancelationHelper

campersau committed:
throw from readFunc to make sure it isn't called

campersau committed:
use s instead of state for consistency

campersau commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -133,25 +133,26 @@ private static async Task<WasmFetchResponse> CallFetch(HttpRequestMessage reques
             List<string> headerValues = new List<string>(headerCount);
             JSObject abortController = BrowserHttpInterop.CreateAbortController();
-            CancellationTokenRegistration? abortRegistration = cancellationToken.Register(() =>
+            CancellationTokenRegistration? abortRegistration = cancellationToken.Register(static s =>
```
I noticed that most of the `cancellationToken.Register` usages didn't use the `state`, so I have updated them to use it.

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpInterop.cs:
```diff
@@ -117,30 +123,27 @@ public static unsafe Task<JSObject> Fetch(string uri, string[] headerNames, stri
 
 
-        public static async ValueTask<T> CancelationHelper<T>(Task<T> promise, CancellationToken cancellationToken, JSObject? abortController, JSObject? fetchResponse)
+        public static async ValueTask CancelationHelper(Task promise, CancellationToken cancellationToken, JSObject? fetchResponse = null)
```
I don't know if that's important. Probably not.

pavelsavara commented:
I guess the ball is on my side of the court now, I will do full review early next week. Sorry for delay.

campersau committed:
improve exception handling

pavelsavara commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -148,7 +109,7 @@ export function http_wasm_fetch_bytes(url: string, header_names: string[], heade
 }
 
-export function http_wasm_fetch(url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[], abort_controller: AbortController, body: Uint8Array | ReadableStream | null): ControllablePromise<ResponseExtension> {
+export function http_wasm_fetch(url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[], abort_controller: AbortController, body: Uint8Array | ReadableStream | null): Promise<ResponseExtension> {
```
this should stay `ControllablePromise`

pavelsavara commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -65,80 +64,42 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
-    const cancelable_promise = http_wasm_fetch(url, header_names, header_values, option_names, option_values, abort_controller, body);
-    pull_state.__fetch_promise_control = loaderHelpers.getPromiseController(cancelable_promise);
-    return cancelable_promise;
+export function http_wasm_fetch_stream(url: string, header_names: string[], header_values: string[], option_names: string[], option_values: any[], abort_controller: AbortController, body: TransformStreamExtension): Promise<ResponseExtension> {
```
return `ControllablePromise`

pavelsavara commented on src/mono/wasm/runtime/http.ts:
```diff
@@ -65,80 +64,42 @@ export function http_wasm_abort_response(res: ResponseExtension): void {
-    pull_state.dispose();
-    controller.error(error);
-    pull_state.__pull_promise_control = null;
+export function http_wasm_transform_stream_write(ts: TransformStreamExtension, bufferPtr: VoidPtr, bufferLength: number): Promise<void> {
```
`ControllablePromise`

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -220,51 +221,71 @@ private static async Task<WasmFetchResponse> CallFetch(HttpRequestMessage reques
                         cancellationToken.ThrowIfCancellationRequested();
 
-                        promise = BrowserHttpInterop.Fetch(uri, headerNames.ToArray(), headerValues.ToArray(), optionNames, optionValues, abortController, buffer);
+                        Task<JSObject> fetchPromise = BrowserHttpInterop.Fetch(uri, headerNames.ToArray(), headerValues.ToArray(), optionNames, optionValues, abortController, buffer);
```
did we lose ability to abort request here ? I mean, not just abandon the promise, but to stop wasting bandwidth.

campersau replied:
No, we still do abort the fetch with the `abortController`.
See https://github.com/dotnet/runtime/pull/91699/files#diff-66ce677ded58c3a6a0639d321f4460bb0b2c4bcaeb3216ff3db4381937365bd8R136

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -329,41 +350,82 @@ static async Task<HttpResponseMessage> Impl(HttpRequestMessage request, Cancella
-                ReadableStreamControllerEnqueueUnsafe(this, _buffer, length);
-            }
-            catch (Exception ex)
+            cancellationToken.ThrowIfCancellationRequested();
```
since we switched from pull to push model, this call could come from any thread. And so it will need to be dispatch to the thread/worker which has the JS objects on it.

campersau replied:
At this point we are only doing two `abortController.Dispose` and `transformStream.Dispose` calls which already dispatch to their `SynchronisationContext`. Also we do use `ConfigureAwait(true)`. Or do will still need to dispatch? Is it only required for `FEATURE_WASM_THREADS`?

E.g.
```c#
#if FEATURE_WASM_THREADS
 transformStream.SynchronizationContext.Send(...);
```

pavelsavara replied:
I mean `TransformStreamWriteUnsafe` is pushing bytes, so wrapping it with `FEATURE_WASM_THREADS` and dispatch should be necessary I think. 

The `FEATURE_WASM_THREADS` is the multi-threaded build of runtime. 
You can build it with `/p:MonoWasmBuildVariant=multithread` 
Make sure to `rm -rf artifacts` before doing so.

pavelsavara replied:
Or you can keep streaming not implemented in multi-threaded build and return false for `SupportsStreamingRequest` under `FEATURE_WASM_THREADS`

pavelsavara replied:
and throw here, just to be sure.

campersau replied:
> I mean TransformStreamWriteUnsafe is pushing bytes, so wrapping it with FEATURE_WASM_THREADS and dispatch should be necessary I think.

Sorry, I was looking at the wrong code.
Yes, it is indeed needed and with https://github.com/dotnet/runtime/pull/91699/commits/0f2ee8f47beb7d92a015c9e3b15489c6d45670d5 the tests passed when building / running with `/p:MonoWasmBuildVariant=multithread` otherwise they would fail.

pavelsavara commented on src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs:
```diff
@@ -220,51 +221,71 @@ private static async Task<WasmFetchResponse> CallFetch(HttpRequestMessage reques
+                                    }
+                                    try
+                                    {
+                                        using (fetchResponse = await fetchTask.ConfigureAwait(true)) // observe exception
```
Could you please explain, I don't follow fully.

campersau replied:
This will observe the abort / fetch exception from fetch otherwise this would throw an `UnobservedTaskException`.
The `using` and `BrowserHttpInterop.AbortResponse(fetchResponse)` part isn't really needed because this will always throw.

pavelsavara replied:
Thanks. Is that strategy valid for all types of exceptions here ? Or just cancellation ?

campersau replied:
Yes, this is valid for all types of exceptions.

campersau committed:
return ControllablePromise

campersau committed:
dispatch TransformStreamWrite on FEATURE_WASM_THREADS

pavelsavara commented:
/azp run runtime-wasm
